drv/psc-psu-update/src/main.rs

```diff
@@ -0,0 +1,262 @@
+
+impl Ticks {
+    fn now() -> Self {
+        Self(sys_get_timer().now)
+    }
+}
+
+impl Add for Ticks {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Self(self.0 + other.0)
+    }
+}
+
+#[derive(Copy, Clone, Default)]
+struct Psu {
+    /// When did we last check this device?
+    last_checked: Option<Ticks>,
+
+    /// Is the device physically present?
+    present: Option<bool>,
+
+    /// Is the device on and with POWER_GOOD set?
+    power_good: Option<bool>,
+
+    /// The last firmware revision read
+    firmware_revision: Option<Mwocp68FirmwareRev>,
+
+    /// Does the firmware we have match the firmware here?
+    firmware_matches: Option<bool>,
+
+    /// What time did we start an update?
+    update_started: Option<Ticks>,
+
+    /// What time did the update complete?
+    update_succeeded: Option<Ticks>,
+
+    /// What time did the update last fail, if any?
+    update_failure: Option<(Ticks, Option<UpdateState>, Mwocp68Error)>,
+
+    /// How long should the next update backoff, if at all? (In ticks.)
+    update_backoff: Option<Ticks>,
+}
+
+impl Psu {
+    fn update_should_be_attempted(&mut self, dev: &Mwocp68, ndx: u8) -> bool {
+        let now = Ticks::now();
+
+        self.last_checked = Some(now);
+        self.power_good = None;
+        self.firmware_matches = None;
+        self.firmware_revision = None;
+
+        if !dev.present() {
+            self.present = Some(false);
+
+            //
+            // If we are seeing our device as not present, we will clear our
+            // backoff value: if/when a PSU is plugged back in, we want to
+            // attempt to update it immediately if the firmware revision
+            // doesn't match our payload.
+            //
+            self.update_backoff = None;
+            return false;
+        }
+
+        self.present = Some(true);
+
+        match dev.power_good() {
+            Ok(power_good) => {
+                self.power_good = Some(power_good);
+
+                if !power_good {
+                    return false;
+                }
+            }
+            Err(err) => {
+                ringbuf_entry!(Trace::PowerGoodFailed(ndx, err));
+                return false;
+            }
+        }
+
+        match dev.firmware_revision() {
+            Ok(revision) => {
+                self.firmware_revision = Some(revision);
+
+                if revision == MWOCP68_FIRMWARE_REV {
+                    self.firmware_matches = Some(true);
+                    return false;
+                }
+
+                self.firmware_matches = Some(false);
+            }
+            Err(err) => {
+                ringbuf_entry!(Trace::FirmwareRevFailed(ndx, err));
+                return false;
+            }
+        }
+
+        if let (Some(started), Some(backoff)) =
+            (self.update_started, self.update_backoff)
+        {
+            if started + backoff > now {
+                ringbuf_entry!(Trace::BackingOff(ndx));
```

Comment:
i wonder if we might want to include the number of remaining ticks to back off for in the ringbuf entry? i don't _think_ that would make the entry enum bigger, given that we have other variants that include an index and error code?

Reply:
I want to avoid flooding the ringbuf with entries (the remaining ticks will change over time). Will add a comment to that effect.

Reply:
Good point, thanks for adding the comment!